syntax = "proto3";

package dfs;

option go_package = "dfs";

message NodeDescriptor {
  string node_id = 1;
  string host = 2;
  int32 grpc_port = 3;
  int64 capacity_bytes = 4;
  int64 free_bytes = 5;
  string mac = 6;
}

message RegisterNodeRequest { NodeDescriptor node = 1; }
message RegisterNodeResponse { bool ok = 1; string reason = 2; }

message HeartbeatRequest {
  string node_id = 1;
  double load_factor = 2;
  int64 free_bytes = 3;
}
message RebalanceInstruction {
  string chunk_id = 1;
  string source_node_id = 2;
  string target_node_id = 3;
}
message HeartbeatResponse {
  bool ok = 1;
  repeated RebalanceInstruction rebalances = 2;
}

message UploadPlanRequest {
  string file_name = 1;
  int64 file_size = 2;
  int64 chunk_size = 3; // requested chunk size in bytes
}
message ChunkPlacement {
  string chunk_id = 1;
  int32 chunk_index = 2;
  repeated NodeDescriptor replicas = 3;
}
message UploadPlanResponse {
  repeated ChunkPlacement placements = 1;
  int64 chunk_size = 2; // actual chunk size master chose
  int32 replication_factor = 3;
}

message ReportChunkStoredRequest {
  string file_id = 1;
  string chunk_id = 2;
  int32 chunk_index = 3;
  string node_id = 4;
}
message ReportChunkStoredResponse { bool ok = 1; }

message FileMetadataRequest { string file_id = 1; }
message FileMetadataResponse {
  string file_id = 1;
  string file_name = 2;
  int64 file_size = 3;
  int64 chunk_size = 4;
  repeated ChunkPlacement placements = 5;
}

message UploadChunkRequest {
  string file_id = 1;
  string chunk_id = 2;
  int32 chunk_index = 3;
  bytes data = 4;
}
message UploadChunkResponse { bool ok = 1; string reason = 2; }

message DownloadChunkRequest { string chunk_id = 1; }
message DownloadChunkResponse {
  bool ok = 1;
  bytes data = 2;
  string reason = 3;
}

message DeleteChunkRequest { string chunk_id = 1; }
message DeleteChunkResponse { bool ok = 1; string reason = 2; }

message HealthCheckRequest {}
message HealthCheckResponse { bool ok = 1; string status = 2; }

service MasterService {
  rpc RegisterNode(RegisterNodeRequest) returns (RegisterNodeResponse);
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
  rpc GetUploadPlan(UploadPlanRequest) returns (UploadPlanResponse);
  rpc ReportChunkStored(ReportChunkStoredRequest) returns (ReportChunkStoredResponse);
  rpc GetFileMetadata(FileMetadataRequest) returns (FileMetadataResponse);
}

service StorageService {
  rpc UploadChunk(UploadChunkRequest) returns (UploadChunkResponse);
  rpc DownloadChunk(DownloadChunkRequest) returns (DownloadChunkResponse);
  rpc DeleteChunk(DeleteChunkRequest) returns (DeleteChunkResponse);
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}
